src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -17,6 +16,258 @@ namespace Microsoft.Quantum.QsCompiler.Transformations.CallGraphWalker
+        /// any duplicates and the last node in the list is assumed to be connected
+        /// to the first.
+        /// </summary>
+        public List<List<int>> GetAllCycles(Dictionary<int, List<int>> graph)
+        {
+            // Possible Optimization: Each SCC could be processed in parallel.
+
+            // Possible Optimization: Getting the subgraph for an SCC only needs
+            // to consider the parent graph, not the full graph. A recursive strategy
+            // may be better than using a stack object.
+
+            var cycles = new List<List<int>>();
+
+            PushSCCs(graph);
+            while (SccStack.Any())
+            {
+                var (scc, startNode) = SccStack.Pop();
+                var subGraph = GetSubGraph(graph, scc);
+                cycles.AddRange(GetSccCycles(subGraph, startNode));
+
+                subGraph.Remove(startNode);
+                foreach (var (_, children) in subGraph)
+                {
+                    children.Remove(startNode);
+                }
+
+                PushSCCs(subGraph);
+            }
+
+            return cycles;
+        }
+
+        /// <summary>
+        /// Uses Tarjan's algorithm for finding strongly-connected components, or SCCs of a
+        /// graph to get all SCC, order them by their minimum-valued node, and push them to
+        /// the SCC stack.
+        /// </summary>
+        private void PushSCCs(Dictionary<int, List<int>> graph)
+        {
+            var sccs = TarjanSCC(graph).OrderByDescending(x => x.MinNode);
+            foreach (var scc in sccs)
+            {
+                SccStack.Push(scc);
+            }
+        }
+
+        private Dictionary<int, List<int>> GetSubGraph(Dictionary<int, List<int>> inputGraph, HashSet<int> subGraphNodes) =>
+            inputGraph
+                .Where(kvp => subGraphNodes.Contains(kvp.Key))
+                .ToDictionary(kvp => kvp.Key, kvp => kvp.Value.Where(dep => subGraphNodes.Contains(dep)).ToList());
+
+        /// <summary>
+        /// Tarjan's algorithm for finding all strongly-connected components in a graph.
+        /// A strongly-connected component, or SCC, is a subgraph in which all nodes can reach
+        /// all other nodes in the subgraph.
+        ///
+        /// This implementation was based on the pseudo-code found here:
+        /// https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+        ///
+        /// This returns a list of SCCs represented by sets of nodes, each paired with the
+        /// value of their lowest valued-node. The list is sorted such that each SCC comes
+        /// before any of its children (reverse topological ordering).
+        /// </summary>
+        private List<(HashSet<int> SCC, int MinNode)> TarjanSCC(Dictionary<int, List<int>> inputGraph)
+        {
+            var index = 0; // This algorithm needs its own separate indexing of nodes
```

Comment:
This comment isn't very useful. I'd say something along the lines, that the index represents the order in which nodes are discovered by the algorithm.

Reply:
When writing this, I though I could be clever about the indexing and use the node's ids, but I discovered that will not work. I wanted to put something here to that effect, in case someone else had the same idea.